### ISSUE TITLE
Clipped video output when window resolution is larger than monitor resolution

### ISSUE THREAD

**Issue by Zinfidel** [mkdasher/mupen64-rr-lua-#22]
### Summary
When video plugin settings are set to render the game at a resolution that is larger than the current resolution of the monitor, the output video is clipped. For example, when using the GLideN64 plugin with a windowed resolution of 2880x2160, on a monitor set to a resolution of 1920x1200, the output video looks like
![clip](https://user-images.githubusercontent.com/1929934/104871191-46a42780-58ff-11eb-89b8-41d5001e01b5.png)
(This screenshot was scaled down to 640x480 post capture to avoid posting a 4k picture to this issue)

In previous versions of mupen (rr-v8), instead of a black area where the video is "clipped" there would be a video "smear" effect where the outermost captured pixel gets repeated to the edge of the video frame. This is an issue for a user that wants to dump a movie at a very high resolution without necessarily having access to a monitor with a very large resolution. This would happen for publications on tasvideos.org, for example.
### Environment
OS: Windows 10 Home
Codec: VFW (Lagarith)
### Expected Behavior
Video is captured at the resolution of the window regardless of the current monitor's setting, or where the emulator's window is on the screen.
### Reproduction Steps
1. Choose a video plugin that allows for arbitrary custom window resolutions
2. Set the window size to a resolution that is larger than the monitor the emulator is displayed on
3. Start a game
4. Start dumping a/v
5. Capture some frames, then stop dumping
6. Examine the video and notice that the video is clipped

**Comment by Madghostek** [comment]
I tried multiple video plugins, and I didn't manage to get any to extend the window past screen size (which can be done, but on plugin's side I think, but I'll look into that), which causes plugin to render the goal size, but it gets clipped when displayed on window... and mupen records that window. In my case I get the smearing mentioned.

Another side effect of recording window instead of collaborating with gfx plugin is fact that gfx or system doesn't care to render offscreen parts at all, which can be proven by displaying the buffer on another window (or, surprise surprise, capturing to avi), which reveals it still holds the old data and it's just not being updated, kind of like this: 
![screen](https://user-images.githubusercontent.com/28148202/105069770-af67dd00-5a82-11eb-9f93-248dbb30aca3.png)
this only happens on jabo,gln64 and direct64 (however in the past I recorded with direct64 without issues so there's definitely something going on)

I guess plugins were never meant to work with screen capture on mupen and it either somehow works or it doesn't. Notice how you can screenshoot with any plugin correctly without black parts, because it works in a different way than screen capture. But plugin developers might know more about this.

**Comment by Aurumaker72** [comment]
This is an issue with windows.

**Comment by Zinfidel** [comment]
> This is an issue with windows.

I've updated the issue with an environment section.

**Comment by Madghostek** [comment]
Small update, I found out GlideN64 *does* collaborate with emulator, it implements `readScreen` which grabs the rendered frame internally instead of copying from display, that's why it's possible to record offscreen sometimes! So the only issue is that the buffer size is based on window size, not resolution in plugin settings, which is very likely why there are black parts/repeated pixel outside of window area. I think it will be possible to force emulator not to snap to screen size somehow

**Comment by Madghostek** [comment]
[mupen64_bigwindow.zip](https://github.com/mkdasher/mupen64-rr-lua-/files/5879451/mupen64_bigwindow.zip)

Okay, this is a release build of the newest commit, could you try to encode with huge resolution to see if now it correctly saves to avi file? (I can't test such big resolutions because mupen runs out of memory)

**Comment by Zinfidel** [comment]
> [mupen64_bigwindow.zip](https://github.com/mkdasher/mupen64-rr-lua-/files/5879451/mupen64_bigwindow.zip)
> 
> Okay, this is a release build of the newest commit, could you try to encode with huge resolution to see if now it correctly saves to avi file? (I can't test such big resolutions because mupen runs out of memory) @Zinfidel

Hey @Madghostek , that zip archive is being immediately quarantined by Windows as containing a trojan, "Win32/Wacatac.B!ml" inside of mupen64.exe. I'm guessing that this is probably a false positive, but if the archive is showing up for me as containing a trojan, that's probably an issue that's going to come up when the code is released.

**Comment by Madghostek** [comment]
Weird, I have Windows Defender active and it didn't complain, however virustotal indeed says there are 2 detections https://www.virustotal.com/gui/file/429f4405cd1b74fe059936d0efb8535dcac8b528650b66e457fbfc54c8fe6a99/detection

I've read this particular detection is a common false positive (in the virus signature there's !ml which means it's machine learning match), but I see it's suspicious, not sure what suddenly caused this because 1.0.5 says no detections... I will try to find commit which caused it, maybe it can be fixed

**Comment by Aurumaker72** [comment]
It's fixed if you use @Madghostek's [gln64 fork](https://github.com/Madghostek/gln64-avi)

**Comment by Aurumaker72** [comment]
Ok nvm now its truly fixed

**Comment by Zinfidel** [comment]
I'm really sorry for how late this response is @Madghostek and @Aurumaker72 . I tested both the executable you posted in this issue, @Madghostek , and I also built the latest commit as of today and tested that. Both are correctly capturing video at high resolutions that exceed the resolution of the monitor that mupen64 is running on. I tested 4K output resolution on a 1080p monitor, and 4K resolution on a 4K monitor but with the window off the screen.

Success!
